### ISSUE TITLE
 skel_vector_loss_layer.cpp:24 Check failed: this->layer_param_.has_skel_vector_param()

### ISSUE THREAD

**Issue by zhangwangzi2010** [chanyn/3Dpose_ssl#17]
When trying to execute the script "sh train.sh“ inside caffe-3dssl/examples/3D_to_2D folder， such error appears as follows.

I0329 16:58:35.306548 24296 layer_factory.hpp:77] Creating layer 3Dboneloss_h36m1
I0329 16:58:35.306560 24296 net.cpp:101] Creating Layer 3Dboneloss_h36m1
I0329 16:58:35.306565 24296 net.cpp:437] 3Dboneloss_h36m1 <- 3dpredict_h36m_slice_pred_0_split_2
I0329 16:58:35.306572 24296 net.cpp:437] 3Dboneloss_h36m1 <- label3d_slice2_1_split_2
I0329 16:58:35.306579 24296 net.cpp:411] 3Dboneloss_h36m1 -> 3Dboneloss_h36m1
this->layer_param_.has_skel_vector_param() = 0
F0329 16:58:35.306607 24296 skel_vector_loss_layer.cpp:24] **Check failed: this->layer_param_.has_skel_vector_param()**
*** Check failure stack trace: ***
    @     0x7fbdee8f15cd  google::LogMessage::Fail()
    @     0x7fbdee8f3433  google::LogMessage::SendToLog()
    @     0x7fbdee8f115b  google::LogMessage::Flush()
    @     0x7fbdee8f3e1e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7fbdeefeeafb  caffe::SkelVectorLossLayer<>::LayerSetUp()
    @     0x7fbdef00cc8d  caffe::Net<>::Init()
    @     0x7fbdef00e6e1  caffe::Net<>::Net()
    @     0x7fbdef16497a  caffe::Solver<>::InitTrainNet()
    @     0x7fbdef164ee7  caffe::Solver<>::Init()
    @     0x7fbdef16527a  caffe::Solver<>::Solver()
    @     0x7fbdef17a143  caffe::Creator_AdamSolver<>()
    @           0x40b1f8  train()
    @           0x407a14  main
    @     0x7fbded890830  __libc_start_main
    @           0x408329  _start
    @              (nil)  (unknown)
Aborted (core dumped)

 It seems that there is something wrong with the configuration with "skel_vector_param" parameter which is decribed in the caffe/src/caffe.proto.
message SkelVectorParameter{
   optional int32 dim = 1 [default = 3];
 }

in the network proto file "examples/3D_to_2D/addsrr.prototxt", the layer description is listed like 
 layer {
   name: "3Dboneloss_h36m1"
   type: "SkelVectorLoss"
   bottom: "3dpredict_h36m"
   bottom: "label3d"
   top: "3Dboneloss_h36m1"
   include {
     phase: TRAIN
   }
 }

do we need to add "skel_vector_param" into the "3Dboneloss_h36m1" layer defintion like this?
 layer {
   name: "3Dboneloss_h36m1"
   type: "SkelVectorLoss"
   bottom: "3dpredict_h36m"
   bottom: "label3d"
   top: "3Dboneloss_h36m1"
   _**skel_vector_param{
    dim : 3
   }**_
   include {
     phase: TRAIN
   }
 }

Thanks for any information

**Comment by zhangwangzi2010** [comment]
we can see that from the file caffe.proto, there is only two values "1" and "3" which can be specified to the "dim" value of skel_vector_param

**Comment by zhangwangzi2010** [comment]
From the following source codes， which value of dim in the skel_vector_param parameter scope should be configured， 1 or 3？

template<typename Dtype>
void SkelVectorLossLayer<Dtype>::LayerSetUp(const vector<Blob<Dtype>*>& bottom, const vector<Blob<Dtype>*>& top)
{  
  LossLayer<Dtype> ::LayerSetUp(bottom,top);
  std::cout <<"this->layer_param_.has_skel_vector_param() = " << this->layer_param_.has_skel_vector_param() << std::endl;
  CHECK(this->layer_param_.has_skel_vector_param());
  **this->dim = this->layer_param_.skel_vector_param().dim();**
}

template <typename Dtype>
void SkelVectorLossLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
    const vector<Blob<Dtype>*>& top) {
  int count = bottom[0]->count();

  const int num = bottom[0] -> num();
  **const int num_skel = bottom[0] -> channels() / this->dim; // for 3d pose**
  // int parent_node[] = {0, 0,1,2, 0,4,5, 0,7,8,9, 8,11,12, 8,14,15 };
  int parent_node[] = {0, 0,1,2, 0,4,5, 0,7,8, 8,10,11, 8,13,14}; 
  const Dtype *blob_data1 = bottom[0] -> cpu_data();
  const Dtype *blob_data2 = bottom[1] -> cpu_data();
  Dtype *bone_data1 = bone_1.mutable_cpu_data();
  Dtype *bone_data2 = bone_2.mutable_cpu_data();
  
  for(int n = 0; n < num; ++n) {
    // except the root node
    for(int c = 0; c < num_skel; ++c){
      for (int d = 0; d < this->dim; ++d) {
        bone_data1[bottom[0]->offset(n,this->dim*c+d)] = blob_data1[bottom[0]->offset(n,this->dim*parent_node[c]+d)] - blob_data1[bottom[0]->offset(n,this->dim*c+d)];
        bone_data2[bottom[1]->offset(n,this->dim*c+d)] = blob_data2[bottom[1]->offset(n,this->dim*parent_node[c]+d)] - blob_data2[bottom[1]->offset(n,this->dim*c+d)];
      }
    //   bone_data1[bottom[0]->offset(n,3*c)] = blob_data1[bottom[0]->offset(n,3*parent_node[c])] - blob_data1[bottom[0]->offset(n,3*c)];
    //   bone_data1[bottom[0]->offset(n,3*c+1)] = blob_data1[bottom[0]->offset(n,3*parent_node[c]+1)] - blob_data1[bottom[0]->offset(n,3*c)+1];
    //   bone_data1[bottom[0]->offset(n,3*c+2)] = blob_data1[bottom[0]->offset(n,3*parent_node[c])+2] - blob_data1[bottom[0]->offset(n,3*c)+2];

    //   bone_data2[bottom[1]->offset(n,3*c)] = blob_data2[bottom[1]->offset(n,3*parent_node[c])] - blob_data2[bottom[1]->offset(n,3*c)];
    //   bone_data2[bottom[1]->offset(n,3*c+1)] = blob_data2[bottom[1]->offset(n,3*parent_node[c]+1)] - blob_data2[bottom[1]->offset(n,3*c)+1];
    //   bone_data2[bottom[1]->offset(n,3*c+2)] = blob_data2[bottom[1]->offset(n,3*parent_node[c])+2] - blob_data2[bottom[1]->offset(n,3*c)+2];
    }
  }
  caffe_sub(
      count,
      bone_1.cpu_data(),
      bone_2.cpu_data(),
      diff_.mutable_cpu_data());
  Dtype dot = caffe_cpu_dot(count, diff_.cpu_data(), diff_.cpu_data());
  Dtype loss = dot / bottom[0]->num() / Dtype(2);
  top[0]->mutable_cpu_data()[0] = loss;
}

**Comment by chanyn** [comment]
Yes, you should add the "dim" value of skel_vector_param, and dim = 2/3.
This layer also computes the bone loss of 2d key points.
We will update prototxt to fix bug.
